clients/client-java/src/main/java/com/datastrato/gravitino/client/KerberosTokenProvider.java

```diff
@@ -0,0 +1,237 @@
+      throw new IOException("Fail to close login context", le);
+    }
+  }
+
+  void setHost(String host) {
+    this.host = host;
+  }
+
+  private static ThreadFactory getThreadFactory(String factoryName) {
+    return new ThreadFactoryBuilder().setDaemon(true).setNameFormat(factoryName + "-%d").build();
+  }
+
+  /**
+   * Creates a new instance of the KerberosTokenProvider.Builder
+   *
+   * @return A new instance of KerberosTokenProvider.Builder
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private String clientPrincipal;
+    private File keyTabFile;
+    private long reloginIntervalSec = 5;
+
+    /**
+     * Sets the client principal for the HTTP token requests.
+     *
+     * @param clientPrincipal The client principal for the HTTP token requests.
+     * @return This Builder instance for method chaining.
+     */
+    public Builder withClientPrincipal(String clientPrincipal) {
+      this.clientPrincipal = clientPrincipal;
+      return this;
+    }
+
+    /**
+     * Sets the keyTabFile for the HTTP token requests.
+     *
+     * @param file The keyTabFile for the HTTP token requests.
+     * @return This Builder instance for method chaining.
+     */
+    public Builder withKeyTabFile(File file) {
+      this.keyTabFile = file;
+      return this;
+    }
+
+    /**
+     * Kerberos need to re-login to keep credential active periodically.
+     *
+     * <p>Sets the re-login interval for KerberosTokenProvider. Default value is 5s.
+     *
+     * @param reloginIntervalSec The Kerberos re-login interval.
+     * @return This Builder instance for method chaining.
+     */
+    public Builder withReloginIntervalSec(long reloginIntervalSec) {
+      this.reloginIntervalSec = reloginIntervalSec;
+      return this;
+    }
+
+    /**
+     * Builds the instance of the KerberosTokenProvider.
+     *
+     * @return The built KerberosTokenProvider instance.
+     */
+    public KerberosTokenProvider build() {
+      KerberosTokenProvider provider = new KerberosTokenProvider();
+
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(clientPrincipal),
+          "KerberosTokenProvider must set clientPrincipal");
+      Preconditions.checkArgument(
+          clientPrincipal.split("@").length == 2, "Principal has the wrong format");
+      provider.clientPrincipal = clientPrincipal;
+
+      if (keyTabFile != null) {
+        Preconditions.checkArgument(
+            keyTabFile.exists(), "KerberosTokenProvider's keytabFile doesn't exist");
+        Preconditions.checkArgument(
+            keyTabFile.canRead(), "KerberosTokenProvider's keytabFile can't read");
+        provider.keytabFile = keyTabFile.getAbsolutePath();
+      }
+
+      try {
+        provider.refreshLoginContext();
+      } catch (LoginException le) {
+        throw new IllegalStateException("Fail to login", le);
+      }
+
+      provider.reloginIntervalSec = reloginIntervalSec;
```

Comment:
I think it is too frequent to do login. 5 secs is too short. As I remembered a TGT will typically be alive for hours.

Also can you get the expire date in in TGT, you can calculate the next login time based the expire date.

Reply:
Fixed.